contrib/babelfishpg_tsql/src/hooks.c

```diff
@@ -5278,3 +5289,85 @@ is_partitioned_table_reloptions_allowed(Datum reloptions)
 	}
 	return true;
 }
+
+static bool
+is_babelfish_builtin_type(Oid typid)
+{
+	bool res = false;
+	HeapTuple	tp;
+	tp = SearchSysCache1(TYPEOID, ObjectIdGetDatum(typid));
+	if (HeapTupleIsValid(tp))
+	{
+		Form_pg_type typtup = (Form_pg_type) GETSTRUCT(tp);
+		res = pg_strcasecmp(get_namespace_name(typtup->typnamespace), "sys") == 0;
+		ReleaseSysCache(tp);
+	}
+	return res;
+}
+
+/*
+ * set_param_collation - sets the collation of the given parameter
+ * 					   based on the sql dialect.
+ *
+ * 	@param param - parameter to set the collation for
+ * 	@return - collation of the parameter
+ */
+static Oid
+set_param_collation(Param *param)
+{
+	/*
+	 * If sql_dialect is PG then we need to set DEFAULT_COLLATION_OID for any param
+	 * to handle special cases such as checking foreign key when tupe is being inserted
+	 * in the table through TDS endpoint.
+	 */
+	if (sql_dialect == SQL_DIALECT_PG && is_babelfish_builtin_type(param->paramtype))
+	{
+		return DEFAULT_COLLATION_OID;
+	}
+	else
+	{
+		return get_typcollation(param->paramtype);
+	}
+}
+
+/*
+ * default_collation_for_builtin_type - returns the default collation for the given
+ * 									   builtin type.
+ *
+ * 	@param typ - given type (as type struct)
+ * 	@return - default collation for the given builtin type based on dialect
+ */
+static Oid
+default_collation_for_builtin_type(Type typ, bool handle_pg_type)
+{
+	Form_pg_type	typtup;
+	Oid				oid = InvalidOid;
+
+	typtup = (Form_pg_type) GETSTRUCT(typ);
+	if (OidIsValid(typtup->typcollation) &&
+		sql_dialect == SQL_DIALECT_TSQL &&
+		pg_strcasecmp(get_namespace_name(typtup->typnamespace), "sys") == 0)
```

Comment:
We can use cache for sys schema OID here as well.

Reply:
Same as above